### PR TITLE
MOBT-334: Day/Night calculation using bounds for time bounded variables

### PR DESCRIPTION
**Pull request by bayliffe**
## What it does
Modifies the day/night terminator calculation to use time bounds mid-point for diagnostics with time bounds.

If the input cube's time coordinate has time bounds the mid-point of these bounds, rather than the time point, is now used in the calculation of the day/night status of the data. This ensures that should a majority of the period described fall into daylight hours, the data is not classified as falling at night. This is an issue as the time point is always located at the end of the time bounds. Without this approach e.g. weather symbols representing e.g. 20-21Z will be marked as night symbols should the sunset at 20:59 which is not suitable.

## Impacts
This PR includes one change the acceptance test data which is for the global weather symbols (3-hour). The annotated figure below shows the data that has changed.

Where they exist, night codes are one lower than their daytime equivalents. The figure shows that all the changes are +- 1 and around the terminator. The red section shows locations that have been reclassified as day symbols following the change (+1). This is because the terminator has been shifted east from the 21Z location to the midpoint of the bounds (18-21Z), it's location at 19:30Z. Points west of this are forecasting for a period that is more day than night.

labelled_terminator_shift

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Review by mspelman07 — CHANGES_REQUESTED** [review]
A couple of small changes.

File: improver/utilities/solar.py [hunk 1]
```diff
@@ -393,7 +402,11 @@ def process(self, cube: Cube) -> Cube:
 
         modified_masks = iris.cube.CubeList()
         for mask_cube in daynight_mask.slices_over("time"):
-            dtval = mask_cube.coord("time").cell(0).point
+            if mask_cube.coord("time").has_bounds():
+                lb, ub = cube.coord("time").cell(0).bound
+                dtval = lb + (ub - lb) / 2
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/solar.py:407]: Yep, but you can't do that with datetime objects. They can't be added as it is meaningless to add two dates. However, you can subtract them to get the time difference between them and you can add a time difference to a date to get a new date, hence the slightly convoluted form above.

**Review by mspelman07 — DISMISSED** [review]

File: improver/utilities/solar.py [hunk 2]
```diff
@@ -377,6 +377,15 @@ def process(self, cube: Cube) -> Cube:
            dt(2017, 1, 1, 12, 0, 1)   -- +1  --> dt(2017, 1, 1, 12, 0, 2)
            dt(2017, 1, 1, 12, 0, 30)  -- +30 --> dt(2017, 1, 1, 12, 1, 0)
 
+        If the cube's time coordinate has time bounds the mid-point of these
+        bounds rather than the time point is used in the calculation. This
+        ensures that should a majority of the period described fall into
+        daylight hours, the data is not classified as falling at night. This
+        is an issue as the time point is always located at the end of the
+        time bounds. Without this approach weather symbols representing
+        e.g. 20-21Z will be marked as night symbols should the sunset at
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/solar.py:386]: I meant this to be "sun set" but lost a space along the way. I've added it back,

File: improver_tests/wxcode/wxcode/test_WeatherSymbols.py [hunk 3]
```diff
@@ -1430,8 +1430,14 @@ def test_basic(self):
 
     def test_day_night(self):
         """Test process returns the right values for night. """
+        twelve_hours = 3600 * 12
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/wxcode/wxcode/test_WeatherSymbols.py:1433]: Done.

**Review by gavinevans — APPROVED** [review]

**Pull request merged**